core/gui/src/app/workspace/service/compile-workflow/compile-workflow.service.ts

```diff
@@ -0,0 +1,398 @@
+      return undefined;
+    }
+    return this.currentCompilationStateInfo.operatorInputSchemaMap[operatorID];
+  }
+
+  public getPortInputSchema(operatorID: string, portIndex: number): PortInputSchema | undefined {
+    return this.getOperatorInputSchema(operatorID)?.[portIndex];
+  }
+
+  public getOperatorInputAttributeType(
+    operatorID: string,
+    portIndex: number,
+    attributeName: string
+  ): AttributeType | undefined {
+    return this.getPortInputSchema(operatorID, portIndex)?.find(e => e.attributeName === attributeName)?.attributeType;
+  }
+
+  /**
+   * Apply the schema propagation result to an operator.
+   * The schema propagation result contains the input attributes of operators.
+   *
+   * If an operator is not in the result, then:
+   * 1. the operator's input attributes cannot be inferred. In this case, the operator dynamic schema is unchanged.
+   * 2. the operator is a source operator. In this case, we need to fill in the attributes using the selected table.
+   *
+   * @param schemaPropagationResult
+   */
+  private _applySchemaPropagationResult(schemaPropagationResult: { [key: string]: OperatorInputSchema }): void {
+    // for each operator, try to apply schema propagation result
+    Array.from(this.dynamicSchemaService.getDynamicSchemaMap().keys()).forEach(operatorID => {
+      const currentDynamicSchema = this.dynamicSchemaService.getDynamicSchema(operatorID);
+
+      // if operator input attributes are in the result, set them in dynamic schema
+      let newDynamicSchema: OperatorSchema;
+      if (schemaPropagationResult[operatorID]) {
+        newDynamicSchema = WorkflowCompilingService.setOperatorInputAttrs(
+          currentDynamicSchema,
+          schemaPropagationResult[operatorID]
+        );
+      } else {
+        // otherwise, the input attributes of the operator is unknown
+        // if the operator is not a source operator, restore its original schema of input attributes
+        if (currentDynamicSchema.additionalMetadata.inputPorts.length > 0) {
+          newDynamicSchema = WorkflowCompilingService.restoreOperatorInputAttrs(currentDynamicSchema);
+        } else {
+          newDynamicSchema = currentDynamicSchema;
+        }
+      }
+
+      if (!isEqual(currentDynamicSchema, newDynamicSchema)) {
+        this.dynamicSchemaService.setDynamicSchema(operatorID, newDynamicSchema);
+      }
+    });
+  }
+
+  /**
+   * Used for automated propagation of input schema in workflow.
+   *
+   * When users are in the process of building a workflow, Texera can propagate schema forwards so
+   * that users can easily set the properties of the next operator. For eg: If there are two operators Source:Scan and KeywordSearch and
+   * a link is created between them, the attributed of the table selected in Source can be propagated to the KeywordSearch operator.
+   */
+  private invokeWorkflowCompilationAPI(): Observable<WorkflowCompilationResponse> {
```

Comment:
Added. The dependency of `ExecuteWorkflowService` is because a static function `getLogicalPlanRequest` is defined in it. In the future PRs we can move it to `compile-workflow.service.ts` when execution no longer needs to send logical plan request.